client/src/utils/collisions.ts

```diff
@@ -0,0 +1,120 @@
+import { ContextualizedLesson } from '../components/Lessons';
+import { intervalFromLesson } from './TimeUtil';
+
+type CompareKeyFunction<T> = (x: T) => number;
+type CompareFunction<T> = (a: T, b: T) => number;
+const chainComparisons =
+  <T>(funcs: CompareFunction<T>[]): CompareFunction<T> =>
+  (a, b) => {
+    for (const f of funcs) {
+      const res = f(a, b);
+      if (res !== 0) return res;
+    }
+
+    return 0;
+  };
+
+const keyCompare =
+  <T>(key: CompareKeyFunction<T>): CompareFunction<T> =>
+  (a, b) =>
+    key(a) - key(b);
+
+const compareLessonNames = (
+  a: ContextualizedLesson,
+  b: ContextualizedLesson,
+): number => a.title.toLowerCase().localeCompare(b.title.toLowerCase());
+
+enum LessonCollisionTimeType {
+  START = 0,
+  END = 1,
+}
+
+type LessonCollisionTime = {
+  time: Date;
+  lessonIdx: number;
+  type: LessonCollisionTimeType;
+};
+
+const toSortedTimes = (
+  lessons: ContextualizedLesson[],
+): LessonCollisionTime[] =>
+  lessons
+    .map(intervalFromLesson)
+    .flatMap((interval, lessonIdx) => [
+      { time: interval.start, lessonIdx, type: LessonCollisionTimeType.START },
+      { time: interval.end, lessonIdx, type: LessonCollisionTimeType.END },
+    ])
+    .toSorted(
+      chainComparisons([
+        keyCompare(({ time }) => time.getTime()),
+        keyCompare(({ type }) =>
+          // Order start times after end times. This makes lessons which just meet at their end
+          // times still count as not colliding and allows them to break collision blocks.
+          // It's up to discussion and testing whether we want this or not, but in the end it
+          // shouldn't matter too much as most of our blocks still have 15 minute gaps inbetween
+          type === LessonCollisionTimeType.START ? 1 : 0,
```

Comment:
This comes down to a really disgusting scenario, where the two blocks just touch at their endpoints. The way it is written right now, this will be the result:
<img width="325" alt="image" src="https://github.com/user-attachments/assets/900dc00d-f4a5-4089-be1e-9cf3a9fc4a69">

ordering them the other way around actually maybe looks better imo:
<img width="325" alt="image" src="https://github.com/user-attachments/assets/f579dc01-6856-4710-b046-5dd8be74304e">

so let's discuss which of them we prefer...